unix/writev_stubs.c

```diff
@@ -0,0 +1,68 @@
+#define _XOPEN_SOURCE /* IOV_MAX */
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/custom.h>
+#include <caml/callback.h>
+#include <caml/alloc.h>
+#include <caml/unixsupport.h>
+#include <caml/bigarray.h>
+#include <caml/threads.h>
+
+#include <stdio.h>
+#include <errno.h>
+#ifndef _WIN32
+#include <sys/uio.h>
+#include <limits.h>
+#endif
+
+CAMLprim
+    value
+    stub_cstruct_iov_max(value unit)
+{
+  CAMLparam1(unit);
+#ifdef IOV_MAX
+  CAMLreturn(Val_int(IOV_MAX));
+#else
+  CAMLreturn(Val_int(0));
+#endif
+}
+
+CAMLprim value stub_cstruct_writev(value fd, value val_list)
+{
+  CAMLparam2(fd, val_list);
+  CAMLlocal5(next, head, val_buf, val_ofs, val_len);
+  int i;
+#ifdef _WIN32
+  caml_failwith("writev is not supported on Win32");
+#else
+  struct iovec iovec[IOV_MAX];
+  int c_fd = Int_val(fd);
+
+  next = val_list;
+  /* Calculate the length of the val_list */
+  int length = 0;
+  for (next = val_list; next != Val_emptylist; next = Field(next, 1))
+    length++;
+  /* Only copy up to the iovec array size */
+  if (length > IOV_MAX)
+    length = IOV_MAX;
```

Comment:
Silently truncating it might be worse than just letting the syscall failing later on in `uerror` with EINVAL.
You exposed the maximum, if the user goes above he should be punished with an exception imho.
It's a bit harmless for a STREAM that has to handle shortcounts, but a DATAGRAM will send half a packet.
(even if we're all a bit insane because IOV_MAX is huge)

Reply:
We should mostly mirror what writev(2) itself does, and that returns an EINVAL if > IOV_MAX.  So I agree that a uerror should be raised if it's too large, and not a silent truncation.